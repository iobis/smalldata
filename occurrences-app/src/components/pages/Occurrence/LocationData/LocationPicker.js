import 'leaflet/dist/leaflet.css'
import iconRetinaUrl from 'leaflet/dist/images/marker-icon-2x.png'
import iconUrl from 'leaflet/dist/images/marker-icon.png'
import L from 'leaflet'
import PropTypes from 'prop-types'
import React, { useState, useEffect, useRef } from 'react'
import shadowUrlUrl from 'leaflet/dist/images/marker-shadow.png'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Map, TileLayer, Marker, Popup } from 'react-leaflet'
import { useDebounce } from '../../../../hooks/hooks'
import { useTranslation } from 'react-i18next'

// This part of code relates to issue with react-leaflet. Please update the code when issues is fixed. Refer
// to https://github.com/PaulLeCam/react-leaflet/issues/453 for more details
const DefaultIcon = L.icon({
  iconRetinaUrl: iconRetinaUrl,
  iconUrl:       iconUrl,
  shadowUrl:     shadowUrlUrl
})

L.Marker.prototype.options.icon = DefaultIcon

export default function LocationPicker({ onChange }) {
  const { t } = useTranslation()
  const mapRef = useRef()
  const [latitude, setLatitude] = useState(51.505)
  const [longitude, setLongitude] = useState(-0.09)
  const coordinates = [latitude, longitude]
  const [zoom, setZoom] = useState(12)
  const [searchString, setSearchString] = useState('')
  const [suggestions, setSuggestions] = useState([])

  function setMarkerCoordinates(latlng) {
    setLatitude(latlng.lat)
    setLongitude(latlng.lng)
    onChange({ latitude: latlng.lat, longitude: latlng.lng })
  }

  const debouncedSearch = useDebounce(searchString, 500)

  useEffect(() => {
    const fetchSuggestions = async() => {
      const response = await fetch(`https://nominatim.openstreetmap.org/search?format=json&q=${searchString}`)
      const suggestions = await response.json()
      setSuggestions(suggestions)
    }

    if (debouncedSearch) fetchSuggestions()
  }, [debouncedSearch])

  useEffect(() => {
    mapRef.current.leafletElement.invalidateSize(true)
  })

  function handleClearClick() {
    setSuggestions([])
    setSearchString('')
  }

  function handleSearchStringChange(newSearchString) {
    setSearchString(newSearchString)
    if (!newSearchString) setSuggestions([])
  }

  return (
    <div className="location-picker section is-fluid">
      <div className="columns">
        <div className="column is-half">
          <div className="location-search field">
            <div className="control has-icons-left has-icons-right">
              <input
                className="input"
                onChange={(e) => handleSearchStringChange(e.target.value)}
                placeholder={t('occurrenceForm.locationData.locationPicker.searchTip')}
                type="text"
                value={searchString}/>
              <span className="icon is-left">
                <FontAwesomeIcon className="search" icon="search"/>
              </span>
              <span className="clear icon is-small is-right">
                <FontAwesomeIcon className="times-circle" icon="times-circle" onClick={handleClearClick}/>
              </span>
            </div>
          </div>
          <h4 className="result title is-4">{t('occurrenceForm.locationData.locationPicker.resultsTitle')}</h4>
          {suggestions.length > 0
            ? <SuggestionsResult onClick={setMarkerCoordinates} suggestions={suggestions}/>
            : <SuggestionsResultEmpty/>}
        </div>
        <div className="column is-half">
          <Map
            center={coordinates}
            onClick={(e) => setMarkerCoordinates(e.latlng)}
            onZoomEnd={(e) => setZoom(e.target.getZoom())}
            ref={mapRef}
            zoom={zoom}>
            <TileLayer
              attribution="&amp;copy <a href=&quot;http://osm.org/copyright&quot;>OpenStreetMap</a> contributors"
              url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"/>
            <Marker
              draggable={true}
              onDragend={(e) => setMarkerCoordinates(e.target.getLatLng())}
              position={coordinates}>
              <Popup>
                <div>latitude: {latitude}</div>
                <div>longitude: {longitude}</div>
              </Popup>
            </Marker>
          </Map>
        </div>
      </div>
    </div>
  )
}

LocationPicker.propTypes = {
  onChange: PropTypes.func.isRequired
}

function SuggestionsResult({ suggestions, onClick }) {
  const { t } = useTranslation()

  return (
    <table className="table is-fullwidth is-striped is-hoverable">
      <thead>
      <tr>
        <th>{t('common.type')}</th>
        <th>{t('common.name')}</th>
        <th>{t('common.longitude')}</th>
        <th>{t('common.latitude')}</th>
      </tr>
      </thead>
      <tbody>
      {suggestions.map(suggestion => (
        <tr
          className="suggestion-row fieldrow"
          key={suggestion.place_id}
          onClick={() => onClick({ lat: suggestion.lat, lng: suggestion.lon })}>
          <td className="type">
            {suggestion.type}
          </td>
          <td className="name">
            {suggestion.display_name}
          </td>
          <td className="longitude">
            {suggestion.lon}
          </td>
          <td className="latitude">
            {suggestion.lat}
          </td>
        </tr>
      ))}
      </tbody>
    </table>
  )
}

SuggestionsResult.propTypes = {
  onClick:     PropTypes.func.isRequired,
  suggestions: PropTypes.array.isRequired
}

function SuggestionsResultEmpty() {
  const { t } = useTranslation()

  return (
    <>
      <p>{t('occurrenceForm.locationData.locationPicker.emptyResult')}</p>
      <p>{t('occurrenceForm.locationData.locationPicker.emptyResultTip')}</p>
    </>
  )
}
