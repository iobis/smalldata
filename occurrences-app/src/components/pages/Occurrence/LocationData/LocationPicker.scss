.location-picker {
  .result.title {
    margin-bottom: 0.5rem;
  }

  .times-circle {
    cursor: pointer;
    pointer-events: initial;
  }

  .suggestion-row {
    cursor: pointer;
  }

  .leaflet-container {
    height: 500px;
    width: 100%;
  }
}
