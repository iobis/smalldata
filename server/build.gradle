plugins {
  id 'java'
  id 'application'
  id "io.franzbecker.gradle-lombok" version "2.1"
  id 'com.github.johnrengelman.shadow' version '4.0.4'
  id 'org.javamodularity.moduleplugin' version '1.4.0'
  id 'org.beryx.runtime' version '1.1.5'
  id 'checkstyle'
  id 'pmd'
}

repositories {
  mavenLocal()
  jcenter()
}

ext {
  vertxVersion = '3.6.3'
  jacksonVersion = '2.9.8'
  junitJupiterEngineVersion = '5.2.0'
  moduleName = 'obis.smalldata'
}

sourceCompatibility = 10

group = 'org.obis'
version = '1.0.0-SNAPSHOT'

mainClassName = "io.vertx.core.Launcher"

def mainVerticleName = "org.obis.smalldata.Starter"
def watchForChange = 'src/**/*'
def doOnChange = './gradlew classes'

runtime {
  options = ['--strip-debug', '--compress', '2', '--no-header-files', '--no-man-pages']
}

dependencies {
  implementation "io.vertx:vertx-core:$vertxVersion"
  implementation "io.vertx:vertx-web:$vertxVersion"
  implementation "io.vertx:vertx-web-client:$vertxVersion"
  implementation "io.vertx:vertx-mongo-client:$vertxVersion"
  implementation "io.vertx:vertx-web-api-contract:$vertxVersion"

  implementation "com.fasterxml.jackson.dataformat:jackson-dataformat-xml:$jacksonVersion"
  implementation "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:$jacksonVersion"

  implementation "de.flapdoodle.embed:de.flapdoodle.embed.mongo:2.2.0"
  implementation "org.tinylog:tinylog:1.3.5"
  implementation "org.tinylog:slf4j-binding:1.3.5"

  compileOnly "org.projectlombok:lombok:1.18.6"

  testImplementation "io.vertx:vertx-junit5:$vertxVersion"
  testImplementation "io.vertx:vertx-junit5:$vertxVersion"
  testImplementation "com.fasterxml.uuid:java-uuid-generator:3.1.5"
  testImplementation "commons-io:commons-io:2.6"

  testRuntime("org.junit.jupiter:junit-jupiter-engine:$junitJupiterEngineVersion")
}

test {
  useJUnitPlatform()
  testLogging {
    events 'PASSED', 'FAILED', 'SKIPPED'
  }
}

run {
  args = ['run', mainVerticleName, "--redeploy=$watchForChange", "--launcher-class=$mainClassName", "--on-redeploy=$doOnChange"]
}

shadowJar {
  manifest {
    attributes 'Main-Verticle': mainVerticleName
  }
  mergeServiceFiles {
    include 'META-INF/services/io.vertx.core.spi.VerticleFactory'
  }
}

checkstyle {
  ignoreFailures false
  maxErrors 0
  maxWarnings 0
  toolVersion "8.17"
}

pmd {
  consoleOutput true
  ruleSetConfig = resources.text.fromFile("config/pmd/ruleset.xml")
  ruleSets = []
}
