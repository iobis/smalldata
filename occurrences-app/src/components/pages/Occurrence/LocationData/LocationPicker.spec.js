import LocationPicker from './LocationPicker'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { mount } from 'enzyme/build'

jest.useFakeTimers()

describe('LocationPicker', () => {
  let wrapper

  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(() =>
      new Promise((resolve) => {
        resolve({
          json: () => ([{
            MRGID:                      18678,
            gazetteerSource:            'IMIS',
            placeType:                  'City',
            latitude:                   59.883299999999998,
            longitude:                  30.25,
            minLatitude:                null,
            minLongitude:               null,
            maxLatitude:                null,
            maxLongitude:               null,
            precision:                  null,
            preferredGazetteerName:     'Sankt-Petersburg',
            preferredGazetteerNameLang: 'Russian',
            status:                     'standard',
            accepted:                   18678
          }, {
            MRGID:                      48986,
            accepted:                   48986,
            gazetteerSource:            null,
            latitude:                   null,
            longitude:                  null,
            maxLatitude:                null,
            maxLongitude:               null,
            minLatitude:                null,
            minLongitude:               null,
            placeType:                  'Province (administrative)',
            precision:                  null,
            preferredGazetteerName:     'Moscow Oblast',
            preferredGazetteerNameLang: 'English',
            status:                     'standard'
          }, {
            MRGID:                      2551,
            accepted:                   2551,
            gazetteerSource:            'ASFA thesaurus',
            latitude:                   51.33185,
            longitude:                  3.20845,
            maxLatitude:                51.3355,
            maxLongitude:               3.216,
            minLatitude:                51.3282,
            minLongitude:               3.2009,
            placeType:                  'Deelgemeente',
            precision:                  663.3265,
            preferredGazetteerName:     'Zeebrugge',
            preferredGazetteerNameLang: 'Dutch',
            status:                     'standard'
          }])
        })
      })
    )
  })

  afterEach(() => {
    global.fetch.mockRestore()
  })

  it('renders correctly', async() => {
    await act(async() => {
      wrapper = mount(createComponent())
    })
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })

  it('fetches data from api and returns it with callback when changing input field', async() => {
    const onChange = jest.fn()
    await act(async() => {
      wrapper = mount(createComponent({ onChange }))
    })
    wrapper.update()
    expect(wrapper.find('.suggestions-result-empty').exists()).toBe(true)
    expect(wrapper.find('.suggestions-result .suggestion-row')).toHaveLength(0)
    expect(wrapper.find('.search-string.input').prop('value')).toBe('')

    await act(async() => {
      wrapper.find('.search-string').simulate('change', { target: { value: 'St. Petersburg, Russ' } })
      jest.runAllTimers()
    })
    wrapper.update()
    expect(onChange).toHaveBeenCalledTimes(0)
    expect(fetch).toHaveBeenCalledTimes(1)
    expect(fetch).toHaveBeenNthCalledWith(1, 'https://api.obis.org/marineregions/getGazetteerRecordsByName.json/St. Petersburg, Russ/true/false')
    expect(wrapper.find('.suggestions-result-empty').exists()).toBe(false)
    expect(wrapper.find('.suggestions-result .suggestion-row')).toHaveLength(2)
    expect(wrapper.find('.search-string.input').prop('value')).toBe('St. Petersburg, Russ')

    await act(async() => {
      wrapper.find('.suggestions-result .suggestion-row').at(0).simulate('click')
    })
    wrapper.update()
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenNthCalledWith(1, {
      coordinateUncertainty: null,
      latitude:              59.8833,
      longitude:             30.25
    })
    expect(fetch).toHaveBeenCalledTimes(1)
    expect(wrapper.find('.suggestions-result-empty').exists()).toBe(false)
    expect(wrapper.find('.suggestions-result .suggestion-row')).toHaveLength(2)
    expect(wrapper.find('.search-string.input').prop('value')).toBe('St. Petersburg, Russ')

    await act(async() => {
      wrapper.find('.suggestions-result .suggestion-row').at(1).simulate('click')
    })
    wrapper.update()
    expect(onChange).toHaveBeenCalledTimes(2)
    expect(onChange).toHaveBeenNthCalledWith(2, {
      coordinateUncertainty: 663.3265,
      latitude:              51.33185,
      longitude:             3.20845
    })
    expect(fetch).toHaveBeenCalledTimes(1)
    expect(wrapper.find('.suggestions-result-empty').exists()).toBe(false)
    expect(wrapper.find('.suggestions-result .suggestion-row')).toHaveLength(2)
    expect(wrapper.find('.search-string.input').prop('value')).toBe('St. Petersburg, Russ')

    await act(async() => {
      wrapper.find('.times-circle').at(0).simulate('click')
    })
    wrapper.update()
    expect(onChange).toHaveBeenCalledTimes(2)
    expect(fetch).toHaveBeenCalledTimes(1)
    expect(wrapper.find('.suggestions-result-empty').exists()).toBe(true)
    expect(wrapper.find('.suggestions-result .suggestion-row')).toHaveLength(0)
    expect(wrapper.find('.search-string.input').prop('value')).toBe('')
  })

  it('trims value from search input field when user updating it', async() => {
    const onChange = jest.fn()
    await act(async() => {
      wrapper = mount(createComponent({ onChange }))
      wrapper.find('.search-string').simulate('change', { target: { value: '    Brugge    ' } })
      jest.runAllTimers()
    })
    wrapper.update()

    expect(onChange).toHaveBeenCalledTimes(0)
    expect(fetch).toHaveBeenCalledTimes(1)
    expect(fetch).toHaveBeenNthCalledWith(1, 'https://api.obis.org/marineregions/getGazetteerRecordsByName.json/Brugge/true/false')
    expect(wrapper.find('.suggestions-result-empty').exists()).toBe(false)
    expect(wrapper.find('.suggestions-result .suggestion-row')).toHaveLength(2)
    expect(wrapper.find('.search-string.input').prop('value')).toBe('    Brugge    ')
  })
})

function createComponent(props) {
  const defaultProps = {
    onChange: jest.fn()
  }
  return <LocationPicker {...defaultProps} {...props}/>
}
