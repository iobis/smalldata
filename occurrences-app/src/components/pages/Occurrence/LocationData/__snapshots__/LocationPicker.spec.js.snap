// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationPicker renders correctly 1`] = `
<div>
  <div
    class="location-picker section is-fluid"
  >
    <div
      class="columns"
    >
      <div
        class="column is-half"
      >
        <div
          class="location-search field"
        >
          <div
            class="control has-icons-left has-icons-right"
          >
            <input
              class="search-string input"
              placeholder="occurrenceForm.locationData.locationPicker.searchTip"
              type="text"
              value=""
            />
            <span
              class="icon is-left"
            >
              <i
                class="search"
                icon="search"
              />
            </span>
            <span
              class="clear icon is-small is-right"
            >
              <i
                class="times-circle"
                icon="times-circle"
              />
            </span>
          </div>
        </div>
        <h4
          class="result title is-4"
        >
          occurrenceForm.locationData.locationPicker.resultsTitle
        </h4>
        <p>
          occurrenceForm.locationData.locationPicker.emptyResult
        </p>
        <p>
          occurrenceForm.locationData.locationPicker.emptyResultTip
        </p>
      </div>
      <div
        class="column is-half"
      >
        <div
          class="leaflet-container leaflet-grab leaflet-touch-drag"
          style="position: relative;"
          tabindex="0"
        >
          <div
            class="leaflet-pane leaflet-map-pane"
            style="left: 0px; top: 0px;"
          >
            <div
              class="leaflet-pane leaflet-tile-pane"
            >
              <div
                class="leaflet-layer "
                style="z-index: 1;"
              >
                <div
                  class="leaflet-tile-container leaflet-zoom-animated"
                  style="left: 0px; top: 0px;"
                >
                  <img
                    alt=""
                    class="leaflet-tile"
                    role="presentation"
                    src="https://a.tile.openstreetmap.org/12/2046/1362.png"
                    style="width: 256px; height: 256px; left: -250px; top: -18px;"
                  />
                </div>
              </div>
            </div>
            <div
              class="leaflet-pane leaflet-shadow-pane"
            >
              <img
                alt=""
                class="leaflet-marker-shadow leaflet-zoom-hide"
                src="marker-shadow.png"
                style="left: 0px; top: 0px;"
              />
            </div>
            <div
              class="leaflet-pane leaflet-overlay-pane"
            />
            <div
              class="leaflet-pane leaflet-marker-pane"
            >
              <img
                alt=""
                class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive leaflet-marker-draggable"
                src="marker-icon.png"
                style="left: 0px; top: 0px; z-index: 0;"
                tabindex="0"
              />
            </div>
            <div
              class="leaflet-pane leaflet-tooltip-pane"
            />
            <div
              class="leaflet-pane leaflet-popup-pane"
            />
          </div>
          <div
            class="leaflet-control-container"
          >
            <div
              class="leaflet-top leaflet-left"
            >
              <div
                class="leaflet-control-zoom leaflet-bar leaflet-control"
              >
                <a
                  aria-label="Zoom in"
                  class="leaflet-control-zoom-in"
                  href="#"
                  role="button"
                  title="Zoom in"
                >
                  +
                </a>
                <a
                  aria-label="Zoom out"
                  class="leaflet-control-zoom-out"
                  href="#"
                  role="button"
                  title="Zoom out"
                >
                  −
                </a>
              </div>
            </div>
            <div
              class="leaflet-top leaflet-right"
            />
            <div
              class="leaflet-bottom leaflet-left"
            />
            <div
              class="leaflet-bottom leaflet-right"
            >
              <div
                class="leaflet-control-attribution leaflet-control"
              >
                <a
                  href="https://leafletjs.com"
                  title="A JS library for interactive maps"
                >
                  Leaflet
                </a>
                 | © 
                <a
                  href="http://osm.org/copyright"
                >
                  OpenStreetMap
                </a>
                 contributors
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
